www/js/lib/zimArchive.js

```diff
@@ -142,31 +142,71 @@ define(['zimfile', 'zimDirEntry', 'util', 'utf8'],
-     * @param {callbackDirEntryList} callback
+     * @param {Object} search The current state.search object
+     * @param {Integer} resultSize The number of dirEntries to find
+     * @param {callbackDirEntryList} callback The function to call with the result
+     * @param {Boolean} noInterim A flag to prevent callback until all results are ready (used in testing) 
      */
-    ZIMArchive.prototype.findDirEntriesWithPrefix = function(prefix, resultSize, callback) {
+    ZIMArchive.prototype.findDirEntriesWithPrefix = function (search, resultSize, callback, noInterim) {
+        var prefix = search.prefix;
         var that = this;
-        var prefixVariants = util.removeDuplicateStringsInSmallArray([prefix, util.ucFirstLetter(prefix), util.lcFirstLetter(prefix), util.ucEveryFirstLetter(prefix)]);
+        // Establish array of initial values that must be searched first. All of these patterns are generated by the full
+        // search type, and some by basic, but we need the most common patterns to be searched first, as it returns search
+        // results much more quickly if we do this (and the user can click on a result before the rarer patterns complete)
+        // NB duplicates are removed before processing search array
+        var startArray = [];
+        // Ensure a search is done on the string exactly as typed
+        startArray.push(prefix);
+        // Normalize any spacing and make string all lowercase
+        prefix = prefix.replace(/\s+/g, ' ').toLocaleLowerCase();
+        // Add lowercase string with initial uppercase (this is a very common pattern)
+        startArray.push(prefix.replace(/^./, function (m) {
+            return m.toLocaleUpperCase();
+        }));
+        // Get the full array of combinations to check number of combinations
+        var fullCombos = util.removeDuplicateStringsInSmallArray(util.allCaseFirstLetters(prefix, 'full'));
+        // Put cap on exponential number of combinations (five words = 3^5 = 243 combinations)
+        search.type = fullCombos.length < 200 ? 'full' : 'basic';
+        // We have to remove duplicate string combinations because util.allCaseFirstLetters() can return some combinations
+        // where uppercase and lowercase combinations are exactly the same, e.g. where prefix begins with punctuation
+        // or currency signs, for languages without case, or where user-entered case duplicates calculated case
+        var prefixVariants = util.removeDuplicateStringsInSmallArray(
+            startArray.concat(
+                // Get basic combinations first for speed of returning results
+                util.allCaseFirstLetters(prefix).concat(
+                    search.type === 'full' ? fullCombos : []
+                )
+            )
+        );
         var dirEntries = [];
+
         function searchNextVariant() {
+            // If user has initiated a new search, cancel this one
+            if (search.state === 'cancelled') return callback([], search);
```

Comment:
Should be replaced by :
```
            if (search.state === 'cancelled' || search.prefix !== originalPrefix) return callback([], {state : 'cancelled'});
```
so that we also trap the case of a search replaced by another one.
And, instead of passing the current `search` object, we can pass a new one that is indeed cancelled, so that `populateListOfArticles` does nothing.